File: 经典排序/sort.py
# 经典排序
def selection_sort(nums: list[int]):
    """选择排序 O(n²) O(1) 不稳定"""
    n = len(nums)

    # 每轮选出待排序区间最小值与待排序区间第一个索引处的元素交换
    for i in range(n - 1):
        min_val = i
        for j in range(i + 1, n):
            if nums[j] < nums[min_val]:
                min_val = j
        nums[i], nums[min_val] = nums[min_val], nums[i]


def bubble_sort(nums: list[int]):
    """冒泡排序 O(n²) O(1) 稳定"""
    n = len(nums)

    # 每轮把最大的泡泡冒到水面
    for i in range(n - 1, 0, -1):
        flag = False
        for j in range(i):
            if nums[j] > nums[j + 1]:
                nums[j], nums[j + 1] = nums[j + 1], nums[i]
                flag = True
        if not flag:  # 此轮冒泡未交换元素，跳出循环
            break


def insertion_sort(nums: list[int]):
    """插入排序 O(n²) O(1) 稳定"""

    # 初始状态默认第一个元素有序
    # 每次 i 向后遍历相当于向已排序的扑克牌插入新的扑克
    for i in range(1, len(nums)):
        base = nums[i]
        j = i - 1
        while j >= 0 and nums[j] > base:
            nums[j + 1] = nums[j]
            j -= 1
        nums[j + 1] = base


class QuickSort:
    """快速排序 O(nlogn) O(n) 不稳定"""

    def partition(self, nums: list[int], left: int, right: int) -> int:
        """哨兵划分"""
        # 以 nums[left] 为基数
        i, j = left, right
        while i < j:
            while i < j and nums[j] >= nums[left]:
                j -= 1
            while i < j and nums[i] < nums[left]:
                i += 1
            # 元素交换
            nums[i], nums[j] = nums[j], nums[i]
        # 将基准数交换至两子数组的分界线
        nums[i], nums[left] = nums[left], nums[i]
        return i

    def quick_sort(self, nums: list[int], left: int, right: int):
        """快速排序"""
        # 子数组长度为 1 时终止递归
        if left >= right:
            return
        # 哨兵划分
        pivot = self.partition(nums, left, right)
        # 递归左子数组、右子数组
        self.quick_sort(nums, left, pivot - 1)
        self.quick_sort(nums, pivot + 1, right)


def quicksort(nums: list[int]) -> list[int]:
    """快速排序，简单写法"""
    if len(nums) < 2:
        return nums

    pivot = nums[0]
    left, right = [], []

    for val in nums[1:]:
        if val <= pivot:
            left.append(val)
        else:
            right.append(val)

    return quicksort(left) + [pivot] + quicksort(right)


s = QuickSort()
arr = [4, 7, 9, 12, 7, 3, 1, 8]
s.quick_sort(arr, 0, len(arr) - 1)
print(arr)

File: go_version/sort.go
package main

import "slices"

func SelectSort(nums []int) {
	n := len(nums)
	for i := range n - 1 {
		minIdx := i
		for j := i + 1; j < n; j++ {
			if nums[j] < nums[minIdx] {
				minIdx = j
			}
		}
		nums[i], nums[minIdx] = nums[minIdx], nums[i]
	}
}

func BubbleSort(nums []int) {
	n := len(nums)
	var flag bool
	for i := n - 1; i > 0; i-- {
		flag = false
		for j := range i {
			if nums[j] > nums[j+1] {
				nums[j], nums[j+1] = nums[j+1], nums[j]
				flag = true
			}
		}
		if !flag {
			break
		}
	}
}

func QuickSort(nums []int) []int {
	if len(nums) < 2 {
		return nums
	}

	pivot := nums[0]
	var left, right []int

	for _, val := range nums[1:] {
		if val <= pivot {
			left = append(left, val)
		} else {
			right = append(right, val)
		}
	}
	return slices.Concat(QuickSort(left), []int{pivot}, QuickSort(right))
}

func medianOfThree(a, b, c int) int {
	if a > b {
		if b > c {
			return b
		}
		if a > c {
			return c
		}
		return a
	}
	if a > c {
		return a
	}
	if b > c {
		return c
	}
	return b
}

// 三分取中法优化的快速排序
func QuickSort_(nums []int) []int {
	if len(nums) < 2 {
		return nums
	}

	// 三分取中法选择基准值
	mid := len(nums) / 2
	pivot := medianOfThree(nums[0], nums[mid], nums[len(nums)-1])

	// 将基准值移到数组的开头
	nums[0], nums[mid] = nums[mid], nums[0]

	var left, right, equal []int
	for _, val := range nums {
		if val < pivot {
			left = append(left, val)
		} else if val > pivot {
			right = append(right, val)
		} else {
			equal = append(equal, val)
		}
	}

	// 合并时确保将等于基准值的元素添加一次
	return slices.Concat(QuickSort_(left), equal, QuickSort_(right))
}

// 插入排序
func InsertionSort(nums []int) {
	for i := 1; i < len(nums); i++ {
		base := nums[i]
		j := i - 1
		for j >= 0 && nums[j] > base {
			nums[j+1] = nums[j]
			j--
		}
		nums[j+1] = base
	}
}

// 归并排序
func MergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	mid := len(arr) >> 1
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])

	return merge(left, right)
}

// 合并两个有序数组
func merge(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))

	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	// 将剩余的元素追加到结果数组中
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)

	return result
}

// 归并排序链表版
func SortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	mid := getMid(head)
	rightSorted := SortList(mid.Next)
	mid.Next = nil
	leftSorted := SortList(head)
	return mergeTwoList(leftSorted, rightSorted)
}

func getMid(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}
	return slow
}

func mergeTwoList(list1, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	cur := dummy

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			cur.Next = list1
			list1 = list1.Next
		} else {
			cur.Next = list2
			list2 = list2.Next
		}
		cur = cur.Next
	}

	if list1 != nil {
		cur.Next = list1
	}
	if list2 != nil {
		cur.Next = list2
	}
	return dummy.Next
}

type ListNode struct {
	Val  int
	Next *ListNode
}
